llvm/lib/IR/RepoTicket.cpp

```diff
@@ -118,76 +127,122 @@ static GODigestState accumulateGOFinalDigestOrInitialDigestAndGetGODigestState(
 
 // Update the GO's hash value by adding the hash of its dependents.
 template <typename Function>
-static bool updateDigestUseDependenciesAndContributions(
-    const GlobalObject *GO, MD5 &GOHash, GOStateMap &Visited, GOInfoMap &GOIMap,
-    Function AccumulateGODigest) {
+static auto updateDigestUseDependenciesAndContributions(
+    const GlobalObject *GO, GOStateMap &Visited, GOInfoMap &GOIMap,
+    MemoizedHashes &HashTable, Function AccumulateGODigest)
+    -> std::tuple<bool, size_t, DigestType> {
   if (GO->isDeclaration())
-    return false;
+    return std::make_tuple(false, std::numeric_limits<size_t>::max(),
+                           NullDigest);
+
+  const auto NumVisited = Visited.size();
+  LLVM_DEBUG(dbgs() << "Computing hash for \"" << GO->getName() << "\" (#"
+                    << NumVisited << ")\n");
+
+  // If the hash has been memorized, we can return the result immediately.
+  const auto TablePos = HashTable.find(GO);
+  if (TablePos != HashTable.end()) {
+    ++VisitedCachedGOTimes;
+    LLVM_DEBUG(dbgs() << "Returning pre-computed hash for \"" << GO->getName()
+                      << "\"\n");
+    return std::make_tuple(true, NumVisited, TablePos->second);
+  }
+
+  MD5 GOHash = MD5();
 
   bool Inserted;
   typename GOStateMap::const_iterator StateIt;
-  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, Visited.size());
+  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, NumVisited);
   if (!Inserted) {
+    assert(NumVisited > StateIt->second);
     // If GO is visited, use the letter 'R' as the marker and use its state as
     // the value.
-    GOHash.update('R');
-    GOHash.update(StateIt->second);
-    return true;
+    GOHash.update(static_cast<uint8_t>(Tags::Backref));
```

Comment:
Could the enum be typed as uint8_t to avoid these casts?

Reply:
I made a mistake here, changed to `static_cast<char>`.